specification/under_development/profiles/Adaptive_Profile.md

```diff
@@ -0,0 +1,988 @@
+specific backend requires choosing a suitable profile and quantum instruction
+set (QIS). Both can be chosen largely independently, though certain instruction
+sets may be incompatible with this (or other) profile(s). The section on the
+[quantum instruction set](#quantum-instruction-set) defines the requirements for
+a QIS to be compatible with the Adaptive Profile. 
+More information about the role of
+the QIS, recommendations for front- and backend providers, as well as the
+distinction between run-time functions and quantum instructions can be found in
+[this document](../Instruction_Set.md).
+
+**Bullet 2: Measurements** <br/>
+
+The second requirement relieves a restriction from the base profile that qubits
+can only be measured at the end of the program and that no quantum operations
+can be performed on a measured qubit. 
+The only requirement here is that all qubits must be measurable, but
+there are no restrictions on when these measurements are made, and there is
+no restriction necessitating that all qubits be measured. I.E. it's perfectly
+valid to measure a subset of qubits used in the program, not have all measured
+qubits in the output, and perform the measurements at different points.
+
+**Bullet 3: Program output** <br />
+
+There is no change to output specifications with respect to the base profile.
+
+The specification of QIR and all its profiles needs to permit to accurately
+reflect the program intent. This includes being able to define and customize the
+program output. The Base Profile specification hence requires explicitly
+expressing which values/measurements are returned by the program and in which
+order. How to express this is defined in the section on [output
+recording](#output-recording).
+
+While it is sufficient for the QPU to do a final measurement of all qubits in a
+predefined order at the end of the program, only the selected subset will be
+reflected in the produced output schema. A suitable output schema can be 
+generated in a post-processing step after the computation on the quantum
+processor itself has completed; customization of the program output hence does
+not require support on the QPU itself.
+
+The defined [output schemas](../output_schemas/) provide different options for
+how a backend may express the computed value(s). The exact schema can be freely
+chosen by the backend and is identified by a string label in the produced
+schema. Each output schema contains sufficient information to allow quantum
+programming frameworks to generate a user-friendly presentation of the returned
+values in the requested order, such as, e.g., a histogram of all results when
+running the program multiple times.
+
+**Bullet 4: Conditional Forward branching** <br />
+The Adaptive Profile can allow for arbitrary forward branching with the restriction being that
+branch instructions cannot express programs that do not terminate. By default, support for
```

Comment:
Since profile compliant QIR programs are supposed to define a single shot of a quantum program, then by default, I'd want Adaptive Profile programs to 1. terminate and 2. always have a value to indicate whether a shot successfully finished in the case that user programs can have real-time errors. I haven't really heard a use case for allowing non-termination in a quantum shot. Outside of the view of a shot, non-termination becomes useful in an IR but I'm not sold on it being useful within a shot. Repeat until success circuits may have non-deterministic length but you expect the success condition to be achieved eventually.

 If a back-end opts into  backwards branching support, or arbitrary function calls, then I'd say it's up to a back-end to provide real-time support for guaranteeing user program termination whether by a watchdog process or some other means since as mentioned it's not statically enforceable by any tooling.

Reply:
"One thing that isn't clear from reading, and it seems this context also switches depending upon topic, is whether the spec views QIR to be run almost-directly on a QPU, or whether it's executed by a runtime which provides classical support and that calls to a QPU."

In our case the distinction between those two things is small or non-existent unless  'run almost-directly on a QPU' here explicitly means classical code or branching are controlled by an FPGA or more integrated hardware.

I think a lot of the core of the issue from my perspective comes down to whether QIR is intended to cover both the case of a classical co processor driving the creation of quantum kernels and also expressing classical logic to be used within the kernel or whether it _only_ is supposed to specify a quantum kernel/shot and the classical processing within it. I believe I've heard the latter from @bettinaheim and @swernli but I don't remember that being explicitly stated in the base profile spec.

Reply:
In my eyes the two are pretty distinct even at this level:

Bare-metal QPU control: gates or instructions which can be lowered directly into the FPGAs or run on the QPU.
Runtime: arbitrarily complicated classical/quantum code which contains contextual information as well as pre/post-processing done purely classically, likely on a machine that is close-but-not-actually the QPU.

Even at a very basic level you can represent programs that may not have full compatability on a QPU but can be transformed or run in a way that they are - but they need additional runtime support to do so.

QIR itself is already able to represent arbitrarily complicated code since it's just LLVM IR. I've always considered profiles as just putting limitations in place so people can build up simple systems first, evolving towards the far more interesting application: embedding a ton of classical processing/context in the QIR that may be used by any runtime to aggresively optimize whatever algorithm and circuit is being run.

Nothing against the incremental approach and in fact I'm for it. But it's also useful to make sure the foundations don't hobble far more complex use-cases.